com.unity.netcode.gameobjects/Runtime/Transports/UTP/BatchedSendQueue.cs

```diff
@@ -110,23 +137,49 @@ public bool PushMessage(ArraySegment<byte> message)
                 return true;
             }
 
-            // Check if there would be enough room if we moved data at the beginning of m_Data.
-            if (m_Data.Length - TailIndex + HeadIndex >= sizeof(int) + message.Count)
+            // Move the data at the beginning of of m_Data. Either it will leave enough space for
+            // the message, or we'll grow m_Data and will want the data at the beginning anyway.
+            if (HeadIndex > 0 && Length > 0)
             {
-                // Move the data back at the beginning of m_Data.
                 unsafe
                 {
                     UnsafeUtility.MemMove(m_Data.GetUnsafePtr(), (byte*)m_Data.GetUnsafePtr() + HeadIndex, Length);
                 }
 
                 TailIndex = Length;
                 HeadIndex = 0;
+            }
 
+            // If there's enough space left at the end for the message, now is a good time to trim
+            // the capacity of m_Data if it got very large. We define "very large" here as having
+            // more than 75% of m_Data unused after adding the new message.
+            if (m_Data.Length - TailIndex >= sizeof(int) + message.Count)
```

Comment:
Yes, you're correct, the list's length always matches its capacity. I'll add a comment about that since that's a non-obvious but important part of how the code is meant to work.

Originally I had planned on getting rid of `TailIndex` and use the list's `Length` instead, but that turned out to be a bit more complicated in the end. Although now that I look at the documentation for `NativeList`, the `Length` property can be set, so maybe I'll modify the code to do that instead of tracking the tail index.

Reply:
After having tried it, we can't rely on the list's capacity as the capacity of our queue here, because `NativeList` may elect to set the capacity higher than what we tell it to be. For example, if calling `SetCapacity(2044)` the capacity will actually be set to 2048. This could be worked around, but I find it simpler to rely on the length of the list instead.